File: app/src/main/java/com/asfoundation/wallet/service/currencies/LocalCurrencyConversionService.java
package com.asfoundation.wallet.service.currencies;

import com.asf.wallet.BuildConfig;
import com.asfoundation.wallet.entity.ConversionResponseBody;
import com.asfoundation.wallet.ui.iab.FiatValue;
import io.reactivex.Observable;
import io.reactivex.Single;
import java.math.RoundingMode;
import retrofit2.http.GET;
import retrofit2.http.Path;

public class LocalCurrencyConversionService {
  public static final String CONVERSION_HOST = BuildConfig.BASE_HOST;

  private final LocalCurrencyConversionService.TokenToLocalFiatApi tokenToLocalFiatApi;
  private final CurrencyConversionRatesPersistence currencyConversionRatesPersistence;

  public LocalCurrencyConversionService(
      LocalCurrencyConversionService.TokenToLocalFiatApi tokenToLocalFiatApi,
      CurrencyConversionRatesPersistence currencyConversionRatesPersistence) {

    this.tokenToLocalFiatApi = tokenToLocalFiatApi;
    this.currencyConversionRatesPersistence = currencyConversionRatesPersistence;
  }

  public Single<FiatValue> getLocalCurrency() {
    return getAppcToLocalFiat("1.0", 18, false).firstOrError();
  }

  public Observable<FiatValue> getAppcToLocalFiat(String value, int scale, boolean getFromCache) {
    if (getFromCache) {
      return currencyConversionRatesPersistence.getAppcToLocalFiat(value, scale)
          .toObservable();
    }
    return tokenToLocalFiatApi.getValueToLocalFiat(value, "APPC")
        .flatMap(response -> {
          FiatValue convertedValue = new FiatValue(response.getAppcValue()
              .setScale(scale, RoundingMode.FLOOR), response.getCurrency(), response.getSymbol());
          return currencyConversionRatesPersistence.saveRateFromAppcToFiat(value, response.getAppcValue()
              .toString(), response.getCurrency(), response.getSymbol())
              .andThen(Observable.just(convertedValue))
              .onErrorReturn(throwable -> {
                throwable.printStackTrace();
                return convertedValue;
              });
        });
  }

  public Observable<FiatValue> getEtherToLocalFiat(String value, int scale) {
    return tokenToLocalFiatApi.getValueToLocalFiat(value, "ETH")
        .flatMap(response -> {
          FiatValue convertedValue = new FiatValue(response.getAppcValue()
              .setScale(scale, RoundingMode.FLOOR), response.getCurrency(), response.getSymbol());
          return currencyConversionRatesPersistence.saveRateFromEthToFiat(value, response.getAppcValue()
              .toString(), response.getCurrency(), response.getSymbol())
              .andThen(Observable.just(convertedValue))
              .onErrorReturn(throwable -> {
                throwable.printStackTrace();
                return convertedValue;
              });
        });
  }

  public Observable<FiatValue> getLocalToAppc(String currency, String value, int scale) {
    return tokenToLocalFiatApi.convertLocalToAppc(currency, value)
        .map(response -> new FiatValue(response.getAppcValue()
            .setScale(scale, RoundingMode.FLOOR), response.getCurrency(), response.getSymbol()));
  }

  public interface TokenToLocalFiatApi {
    @GET("broker/8.20180518/exchanges/{valueFrom}/convert/{appcValue}")
    Observable<ConversionResponseBody> getValueToLocalFiat(@Path("appcValue") String appcValue,
        @Path("valueFrom") String valueFrom);

    @GET("broker/8.20180518/exchanges/{localCurrency}/convert/{value}?to=APPC")
    Observable<ConversionResponseBody> convertLocalToAppc(@Path("localCurrency") String currency,
        @Path("value") String value);
  }
}

File: app/src/main/java/com/asfoundation/wallet/service/currencies/LocalCurrencyConversionService.kt
package com.asfoundation.wallet.service.currencies

import com.appcoins.wallet.core.network.microservices.api.broker.TokenToLocalFiatApi
import com.appcoins.wallet.core.network.microservices.model.ConversionResponseBody
import com.asfoundation.wallet.ui.iab.FiatValue
import io.reactivex.Single
import java.math.RoundingMode
import javax.inject.Inject

class LocalCurrencyConversionService @Inject constructor(
  private val tokenToLocalFiatApi: TokenToLocalFiatApi,
  private val currencyConversionRatesPersistence: CurrencyConversionRatesPersistence
) {

  val localCurrency: Single<FiatValue>
    get() = getAppcToLocalFiat("1.0", 18)

  fun getAppcToLocalFiat(
    value: String, scale: Int,
    getFromCache: Boolean = false
  ): Single<FiatValue> {
    return if (getFromCache) {
      currencyConversionRatesPersistence.getAppcToLocalFiat(value, scale)
    } else getValueToFiat(value, "APPC", null, scale)
      .flatMap {
          currencyConversionRatesPersistence.saveRateFromAppcToFiat(value, it.amount
              .toString(), it.currency, it.symbol)
              .andThen(Single.just(it))
              .onErrorReturn { throwable: Throwable ->
                throwable.printStackTrace()
                it
              }
        }
  }

  fun getEtherToLocalFiat(value: String, scale: Int): Single<FiatValue> {
    return getValueToFiat(value, "ETH", null, scale)
        .flatMap {
          currencyConversionRatesPersistence.saveRateFromEthToFiat(value, it.amount
              .toString(), it.currency, it.symbol)
              .andThen(Single.just(it))
              .onErrorReturn { throwable: Throwable ->
                throwable.printStackTrace()
                it
              }
        }
  }

  fun getFiatToAppc(currency: String, value: String, scale: Int): Single<FiatValue> {
    return tokenToLocalFiatApi.convertFiatToAppc(currency, value)
        .map { response: ConversionResponseBody ->
          FiatValue(response.value
              .setScale(scale, RoundingMode.FLOOR), response.currency, response.sign)
        }
  }

  fun getValueToFiat(value: String, currency: String, targetCurrency: String? = null,
                     scale: Int): Single<FiatValue> {
    val api = if (targetCurrency != null) tokenToLocalFiatApi.getValueToTargetFiat(currency, value,
        targetCurrency) else tokenToLocalFiatApi.getValueToTargetFiat(currency, value)
    return api.map { response: ConversionResponseBody ->
      FiatValue(response.value
          .setScale(scale, RoundingMode.FLOOR), response.currency, response.sign)
    }
  }

  fun getFiatToLocalFiat(currency: String, value: String, scale: Int): Single<FiatValue> {
    return tokenToLocalFiatApi.getValueToTargetFiat(currency, value)
        .map { response: ConversionResponseBody ->
          FiatValue(response.value
              .setScale(scale, RoundingMode.FLOOR), response.currency, response.sign)
        }
  }
}